File: 2025/01 January/17 Product array puzzle/main.cpp
// Link: https://www.geeksforgeeks.org/problems/product-array-puzzle4525/1

#include <iostream>
#include <vector>
using namespace std;

class Solution
{
public:
    vector<int> productExceptSelf(vector<int> &arr)
    {
        int n = arr.size();
        vector<int> result(n, 1);
        int left = 1, right = 1;
        for (int i = 1; i < n; i++)
        {
            left *= arr[i - 1];
            result[i] *= left;

            right *= arr[n - i];
            result[n - i - 1] *= right;
        }
        return result;
    }
};

int main()
{
    Solution object;
    vector<int> arr = {10, 3, 5, 6, 2};
    vector<int> resule = object.productExceptSelf(arr);
    for (int i = 0; i < resule.size(); i++)
    {
        cout << resule[i] << " ";
    }
    cout << endl;

    arr = {12, 0};
    resule = object.productExceptSelf(arr);
    for (int i = 0; i < resule.size(); i++)
    {
        cout << resule[i] << " ";
    }
    cout << endl;
    return 0;
}

File: 2025/01 January/17 Product array puzzle/main.py
# Link: https://www.geeksforgeeks.org/problems/product-array-puzzle4525/1


class Solution:
    def productExceptSelf(self, arr):
        n = len(arr)
        result = [1] * n
        left = right = 1
        for i in range(n):
            result[i] *= left
            left *= arr[i]
            result[~i] *= right
            right *= arr[~i]
        return result


if __name__ == "__main__":
    obj = Solution()
    arr = [10, 3, 5, 6, 2]
    print(obj.productExceptSelf(arr))

    arr = [12, 0]
    print(obj.productExceptSelf(arr))
